core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -1541,39 +1541,18 @@ private StageId getStageId(PlanFragmentId fragmentId)
 
         private void scheduleTasks()
         {
-            long standardTasksWaitingForNode = preSchedulingTaskContexts.getWaitingForNodeTasksCount(STANDARD);
-            long speculativeTasksWaitingForNode = preSchedulingTaskContexts.getWaitingForNodeTasksCount(SPECULATIVE);
-            long eagerSpeculativeTasksWaitingForNode = preSchedulingTaskContexts.getWaitingForNodeTasksCount(EAGER_SPECULATIVE);
-
-            while (!schedulingQueue.isEmpty()) {
-                PrioritizedScheduledTask scheduledTask;
+            scheduleTasks(EAGER_SPECULATIVE);
+            scheduleTasks(STANDARD);
+            if (preSchedulingTaskContexts.getWaitingForNodeTasksCount(STANDARD) == 0) {
```

Comment:
maybe instead of count we should have hasNodesWaitingForTasks() method?

Reply:
I can change